File: c#/04-InterfaceSegregationPrinciple.cs

// Interface Segregation Principle Violation

interface IWorkable
{
    bool CanCode();
    string Code();
    string Test();
}

class Programmer : IWorkable
{
    public bool CanCode()
    {
        return true;
    }

    public string Code()
    {
        return "Coding";
    }

    public string Test()
    {
        return "Testing in localhost";
    }
}

class Tester : IWorkable
{
    public bool CanCode()
    {
        return false;
    }

    public string Code()
    {
         throw new Exception("Opps! I can not code");
    }

    public string Test()
    {
        return "Testing in test server";
    }
}

class ProjectManagement
{
    public void ProcessCode(IWorkable member)
    {
        if (member.CanCode()) {
            member.Code();
        }
    }
}

// Refactored

interface ICodeable
{
    string Code();
}

interface ITestable
{
    string Test();
}

class Programmer : ICodeable, ITestable
{
    public string Code()
    {
        return "Coding";
    }

    public string Test()
    {
        return "Testing in localhost";
    }
}

class Tester : ITestable
{
    public string Test()
    {
        return "Testing in test server";
    }
}

class ProjectManagement
{
    public void ProcessCode(ICodeable member)
    {
        member.Code();
    }
}

File: ts/04-interfaceSegregationPrinciple.ts

// Interface Segregation Principle Violation

interface IWorkable {

    canCode(): boolean;
    code(): string;
    test(): string;
}

class Programmer implements IWorkable {

    canCode(): boolean {
        return true;
    }

    code(): string {
        return 'Coding';
    }

    test(): string {
        return 'Testing in localhost';
    }
}

class Tester implements IWorkable {

    canCode(): boolean {
        return false;
    }

    code(): string {
        throw new Error('Opps! I can not code');
    }

    test(): string {
        return 'Testing in test server';
    }
}

class ProjectManagement {

    processCode(member: IWorkable): void {

        if (member.canCode()) {
            member.code();
        }
    }
}

// Refactored

interface ICodeable {

    code(): string;
}

interface ITestable {

    test(): string;
}

class Programmer implements ICodeable, ITestable {

    code(): string {
        return 'Coding';
    }

    test(): string {
        return 'Testing in localhost';
    }
}

class Tester implements ITestable {

    test(): string {
        return 'Testing in test server';
    }
}

class ProjectManagement {

    processCode(member: ICodeable): void {
        member.code();
    }
}
